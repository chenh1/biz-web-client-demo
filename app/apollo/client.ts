// import { ApolloLink, HttpLink } from "@apollo/client";
// import { onError } from "@apollo/client/link/error";

// import {
//   NextSSRInMemoryCache,
//   NextSSRApolloClient,
// } from "@apollo/experimental-nextjs-app-support/ssr";
// import { NextSSRInMemoryCache } from "@apollo/experimental-nextjs-app-support/ssr";
// import { registerApolloClient } from "@apollo/experimental-nextjs-app-support";}

// const errorLink = onError(({ graphQLErrors, networkError }) => {
//   if (graphQLErrors)
//     graphQLErrors.forEach(({ message, locations, path, ...rest }) => {
//       Object.keys(rest)?.map(key => console.log(key, rest[key]))
//       console.log(
//         `[GraphQL error]: Message: ${message}, Location: ${locations}, Path: ${path}`
//       )
//     });
//   if (networkError) console.log(`[Network error]: ${networkError}`);
// });

// const httpLink = new HttpLink({
//   uri: `${process.env.API_URL || "http://127.0.0.1:1337"}/graphql`,
// })

// export const { getClient } = registerApolloClient(() => {
//   return new NextSSRApolloClient({
//     cache: new NextSSRInMemoryCache(),
//     link: ApolloLink.from([errorLink, httpLink]),
//   });
// });

import { HttpLink } from "@apollo/client";
import {
  registerApolloClient,
  ApolloClient,
  InMemoryCache,
} from "@apollo/experimental-nextjs-app-support";

export const { getClient, query, PreloadQuery } = registerApolloClient(() => {
  return new ApolloClient({
    cache: new InMemoryCache(),
    link: new HttpLink({
      // this needs to be an absolute url, as relative urls cannot be used in SSR
      uri: `${process.env.API_URL || "http://127.0.0.1:1337"}/graphql`,
      // you can disable result caching here if you want to
      // (this does not work if you are rendering your page with `export const dynamic = "force-static"`)
      // fetchOptions: { cache: "no-store" },
    }),
  });
});