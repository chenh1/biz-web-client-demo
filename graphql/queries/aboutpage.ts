import { gql, DocumentNode } from "@apollo/client";

const aboutPageQuery: DocumentNode = gql`
  query AboutPage {
    aboutPage {
      data {
        attributes {
          title
          content
          topImage {
            data {
              attributes {
                url
                width
                height
                alternativeText
              }
            }
          }
        }
      }
    }
  }
`;

export default aboutPageQuery;