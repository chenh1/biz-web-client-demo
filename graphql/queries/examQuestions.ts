import { gql, DocumentNode } from "@apollo/client";

export type MultiPoints = {
  open: number;
  high: number;
  low: number;
  close: number;
}

export type ChartData = {
  multiPoints: MultiPoints;
  x: number;
  y: number;
}

export type Answer = {
  key: string;
  answer: string;
  value?: string;
  xCell?: number;
  yCell?: number;
  x?: number;
  y?: number;
}

export type Cell = {
  header: string;
  position: number;
}

export type CellValue = {
  value: string | number;
  xCell: number;
  yCell: number;
}

export type Question = {
  question: string;
  correctAnswers: Answer[];
  
}

export type InteractiveTableQuestion = Question & {
  answers: Answer[];
  tableData: CellValue[];
  rowHeaders: Cell[];
  columnHeaders: Cell[];
  answerType: string;
}

export type MultiChoiceQuestion = Question & {
  answers: Answer[];
}

export type InteractiveChartQuestion = Question & {
  xType: string;
  yType: string;
  xAxisLabel: string;
  yAxisLabel: string;
  chartType: "line" | "candlestick";
  chartData: ChartData[];
  interactiveChartData: ChartData[];
  correctAnswer: Answer;
}

export type ExamQuestionType = {
  explanation: string;
  lessonModule: number;
  multiChoiceQuestion?: MultiChoiceQuestion;
  fillInQuestion?: Question;
  interactiveTableQuestion?: InteractiveTableQuestion;
  interactiveChartQuestion?: InteractiveChartQuestion;
  tags: {
    data: {
      attributes: {
        displayName: string;
        color: string;
      }
    }
  }
}

export type ExamQuestionResponse = {
  examQuestions: {
    data: {
      attributes: ExamQuestionType
    }[]
  }
}

export type Datum = {
  [key: string]: number | string;
}

export const examQuestionsQuery: DocumentNode = gql`
  query ExamQuestions($lessonModule: Float) {
    examQuestions(filters: {lessonModule: {eq: $lessonModule }}) {
      data {
        attributes {
          tags {
            data {
              attributes {
                displayName
                color
              }
            }
          }
          explanation
          lessonModule
          multiChoiceQuestion {
            question
            answers {
              answer
              key
            }
            correctAnswers {
              key
            }
          }
          fillInQuestion {
            question
            correctAnswers {
              answer
            }
          }
          interactiveTableQuestion {
            question
            answerType
            tableData(pagination: { page: 1, pageSize: 100 }) {
              value
              xCell
              yCell
            }
            correctAnswers {
              value
              xCell
              yCell
            }
            rowHeaders {
              header
              position
            }
            columnHeaders {
              header
              position
            }
          }
          interactiveChartQuestion {
            xType
            yType
            xAxisLabel
            yAxisLabel
            question
            chartType
            chartData(pagination: { page: 1, pageSize: 100 }) {
              x
              y
              multiPoints
            }
            correctAnswer {
              x
              y
            }
          }
        }
      }
    }
  }
`