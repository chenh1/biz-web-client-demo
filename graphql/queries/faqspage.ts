import { gql, DocumentNode } from "@apollo/client";

const faqsPageQuery: DocumentNode = gql`
  query FaqsPage {
    faqsPage {
      data {
        attributes {
          title
          description
          faq(pagination: { page: 1, pageSize: 50 }) {
            question
            answer
            category {
              data {
                attributes {
                  displayName
                  value
                }
              }
            }
            answerRich
          }
        }
      }
    }
  }
`;

export default faqsPageQuery;