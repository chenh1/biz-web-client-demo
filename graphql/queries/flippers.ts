import { gql, DocumentNode } from "@apollo/client";

export const flippersQuery: DocumentNode = gql`
  query FlipperManager {
    flipperManager {
      data {
        attributes {
          ads
          adsPostsPage
          adsPostPage
          adsDashboard
          adsModuleLearnView
          adsModuleExamView
          testMode
          maintenanceMode
        }
      }
    }
  }
`;
