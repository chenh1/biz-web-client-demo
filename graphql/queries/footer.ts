import { gql, DocumentNode } from "@apollo/client";

export type FooterData = {
  linkGroups: {
    title: string;
    link: {
      url: string;
      copy: string;
    }[]
  }[];
  disclaimer: string;
}

export type FooterDataRes = {
  footer: {
    data: {
      attributes: FooterData
    }
  }
}

export const footerQuery: DocumentNode = gql`
  query Footer {
    footer {
      data {
        attributes {
          linkGroups {
            link {
              url
              copy
            }
            title
          }
          disclaimer
        }
      }
    }
  }
`