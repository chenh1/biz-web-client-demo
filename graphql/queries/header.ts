import { gql, DocumentNode } from "@apollo/client";

export type HeaderData = {
  navLinks: {
    url: string;
    copy: string;
  }[]
}

export type HeaderDataRes = {
  header: {
    data: {
      attributes: HeaderData
    }
  }
}

export const headerQuery: DocumentNode = gql`
  query Header {
    header {
      data {
        attributes {
          navLinks {
            url
            copy
          }
        }
      }
    }
  }
`