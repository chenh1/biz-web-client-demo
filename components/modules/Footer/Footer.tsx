"use client";

import type { FooterDataRes } from "@/graphql/queries/footer";
import React from "react"
import { useQuery, ApolloError } from "@apollo/client";
import { Section } from "../../containers/Section/Section";
import { Box } from "../../containers/Box/Box";
import { Heading } from "../../core/Heading/Heading";
import { SpanLink } from "../../core/SpanLink/SpanLink";
import Link from "next/link"
import { Divider } from "../../core/Divider/Divider";
import { Paragraph } from "../../core/Paragraph/Paragraph";
import { footerQuery } from "../../../graphql/queries/footer";
import { Markdown } from "../../utils/Markdown";
import { Brand } from "../../art/Brand/Brand";
//import { useIntersectionObserver } from "../../hooks/useIntersectionObserver";

interface FooterProps {
  brandContent?: React.ReactNode;
  bgColor?: `bg-${string}-${number}`
}

export const Footer: React.FC<FooterProps> = ({ brandContent, bgColor = 'bg-slate-400' }: FooterProps) => {
  const { 
    data,
    loading,
    error 
  }: { data: FooterDataRes | undefined; loading?: boolean; error?: ApolloError; } = useQuery(footerQuery);
  const footerData = data?.footer?.data?.attributes

  return (
    <div className={`relative ${bgColor} w-screen dark max-w-full`}>
      <Section>
        <Box padding="0" gap="lg" className="gap-12 xl:gap-40">
          <Box padding="0" gap="lg" className="xl:grid-cols-4 items-start">
            {brandContent ||
              <Box padding="0">
                <Link href="/">
                  <Brand mono/>
                </Link>
              </Box>
            }
            <Box padding="0" gap="lg" className="grid-cols-2 xl:grid-cols-3 xl:col-span-3 items-start">
              {footerData?.linkGroups?.map(({ title, link }, i) => {
                return (
                  <Box padding="0" gap="xs" key={i}>
                    <Paragraph className="font-bold">
                      {title}
                    </Paragraph>
                    <Divider className="max-w-[200px]" />
                    <ul className="flex flex-col">
                      {link.map(({ url, copy }, i) => (
                        <li key={i} className="inline mt-0 md:mt-2 ml-0">
                          <Link href={url}>
                            <SpanLink uncolored>
                              {copy}
                            </SpanLink>
                          </Link>
                        </li>  
                      ))}
                    </ul>
                  </Box>
                )
              })}
            </Box>
          </Box>
          <Box padding="0" gap="lg">
            <Divider/>
            <Markdown gap="sm" paragraphSize="sm">
              {footerData?.disclaimer ? [footerData?.disclaimer] : []}
            </Markdown>
          </Box>
        </Box>
      </Section>
    </div>
  )
}