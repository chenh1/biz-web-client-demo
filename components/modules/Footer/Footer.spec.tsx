import { Footer } from "./Footer";
import { MockedProvider } from "@apollo/client/testing";
import { render, waitFor } from '@testing-library/react';
import { JSDOM } from "jsdom"
import { footerQuery } from "../../../graphql/queries/footer";
const dom = new JSDOM()
global.document = dom.window.document
global.window = dom.window

jest.mock("react-markdown", () => ({ children }) => <div>{children}</div>);
jest.mock("rehype-raw", () => ({ children }) => <div>{children}</div>);
jest.mock("next/link", () => ({ url, children }) => <a href={url}>children</a>);

const mocks = [
  {
    request: {
      query: footerQuery
    },
    result: {
      data: {
        footer: {
          data: {
            attributes: {
              disclaimer: "Test Disclaimer",
              linkGroups: [
                {
                  title: "Test Title",
                  link: [
                    {
                      copy: "Test Link",
                      url: "https://test.com"
                    },
                    {
                      copy: "Test Link 2",
                      url: "https://test2.com"
                    }
                  ]
                },
                {
                  title: "Test Title 2",
                  link: [
                    {
                      copy: "Test Link 2",
                      url: "https://test2.com"
                    },
                    {
                      copy: "Test Link 3",
                      url: "https://test3.com"
                    }
                  ]
                }
              ]
            }
          }
        }
      }
    }
  }
];


it('renders Footer module', async () => {
  const { container } = render(
    <MockedProvider mocks={mocks}>
      <Footer />
    </MockedProvider>
  );

  await waitFor(() => new Promise((res) => setTimeout(res, 800)));

  expect(container).toMatchSnapshot();
});
