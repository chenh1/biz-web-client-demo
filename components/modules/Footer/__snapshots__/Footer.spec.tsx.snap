// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders Footer module 1`] = `
<div>
  <div
    class="relative bg-slate-400 w-screen dark max-w-full"
  >
    <div
      class=" flex border-box justify-around w-full py-24 "
    >
      <div
        class="w-full max-w-7xl px-4 xl:px-0"
      >
        <div
          class=" grid  gap-12  p-2 gap-12 xl:gap-40"
        >
          <div
            class=" grid  gap-12  p-2 xl:grid-cols-4 items-start"
          >
            <div
              class=" grid  gap-4  p-2 "
            >
              <a>
                children
              </a>
            </div>
            <div
              class=" grid  gap-12  p-2 grid-cols-2 xl:grid-cols-3 xl:col-span-3 items-start"
            >
              <div
                class=" grid  gap-2  p-2 "
              >
                <p
                  class=" text-night-black dark:text-max-white font-body font-body text-base font-bold"
                >
                  Test Title
                </p>
                <hr
                  class="w-full h-0 border-0 border-t-[1px] border-night-black-300 dark:border-max-white max-w-[200px]"
                />
                <ul
                  class="flex flex-col"
                >
                  <li
                    class="inline mt-0 md:mt-2 ml-0"
                  >
                    <a>
                      children
                    </a>
                  </li>
                  <li
                    class="inline mt-0 md:mt-2 ml-0"
                  >
                    <a>
                      children
                    </a>
                  </li>
                </ul>
              </div>
              <div
                class=" grid  gap-2  p-2 "
              >
                <p
                  class=" text-night-black dark:text-max-white font-body font-body text-base font-bold"
                >
                  Test Title 2
                </p>
                <hr
                  class="w-full h-0 border-0 border-t-[1px] border-night-black-300 dark:border-max-white max-w-[200px]"
                />
                <ul
                  class="flex flex-col"
                >
                  <li
                    class="inline mt-0 md:mt-2 ml-0"
                  >
                    <a>
                      children
                    </a>
                  </li>
                  <li
                    class="inline mt-0 md:mt-2 ml-0"
                  >
                    <a>
                      children
                    </a>
                  </li>
                </ul>
              </div>
            </div>
          </div>
          <div
            class=" grid  gap-12  p-2 "
          >
            <hr
              class="w-full h-0 border-0 border-t-[1px] border-night-black-300 dark:border-max-white "
            />
            <div>
              Test Disclaimer
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
